### PR TITLE
Update post about infinite scroller

### PR DESCRIPTION
**Pull request by surma**
r @jeffposnick? @petele?

cc @paullewis @paulirish @ianvollick @flackr

Live preview: http://pr-3043-dot-weasel-dot-web-central.appspot.com/web/updates/2016/07/infinite-scroller?hl=en

### CODE REVIEW

**Comment by surma** [comment]
Live preview at http://pr-3043-dot-weasel-dot-web-central.appspot.com/web/updates

**Comment by jpmedley** [comment]
I'm giving this a once-over.

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 1]
```diff
@@ -0,0 +1,246 @@
+---
+layout: updates/post
+title: "Complexities of an infinite scroller"
+description: "Infinite scrollers are a common UI pattern. We are trying to solve the technical challenges hidden within."
+published_on: 2016-07-01
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: This description is very general. Could you write one that mentions specific PWA-related problems?

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+authors:
+  - surma
+tags:
+  - javascript
+  - css
+featured_image: /web/updates/images/2016/07/infinite-scroller/screenshot.png
+
+---
+
+TL;DR: [Demo][Infscroller Demo] of an infinite scroller, [code][Infscroller Code] for the infinite scroller
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: Here's a page that shows the way to do tl:drs https://github.com/google/WebFundamentals/blob/master/src/content/en/fundamentals/design-and-ui/animations/index.markdown

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+<iframe width="100%" height="315" src="https://www.youtube.com/embed/zp3vDFgxGXI" frameborder="0" allowfullscreen></iframe>
+
+The UI pattern that are infinite scrollers pops up all over the internet. Google Music’s artist list is one, Facebook’s timeline is one and Twitter’s live feed is one as well. You scroll down and before you reach the bottom, new content magically appears seemingly out of nowhere. It’s a seamless experience for users and it’s easy to see the appeal.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: 'The UI pattern that are infinite scrollers' > 'Infinite scrollers'.

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+The technical challenge behind a infinite scroller, however, is harder than it seems. The range of problems you encounter when you want to do The Right Thing™ is vast. It starts with simple things like the links in the footer becoming practically unreachable because content keeps pushing the footer away. But the problems get harder. How do you handle a resize event when someone turns their phone from portrait to landscape or how do you prevent your phone from grinding to a painful halt when the list gets too long?
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: 'a infinite scroller' > 'an infinite scroller'.

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+#The Right Thing™
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: The internal headings need to be ##, not #.

**Comment by surma** [comment]
Awesome, thanks @jpmedley

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+We thought that was reason enough to come up with a reference implementation that shows a way to tackle all these problems in a reusable way while maintaining performance standards.
+
+We are going to utilize 3 techniques to achieve our goal: DOM recycling, tombstones and scroll anchoring.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: 'use', not 'utilize'.

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+<img src="/web/updates/images/2016/07/infinite-scroller/screenshot.png">
+
+#DOM recycling
+DOM recycling is a (heavily underutilized) technique to keep the DOM node count low. Admittedly, DOM nodes themselves are cheap, but they are not free, as each of them adds extra cost in memory, layout, style and paint. Low-end devices will get noticeably slower if not completely unusable if the website has too big of a DOM to manage. Also keep in mind that every relayout and reapplication of your styles – a process that is triggered whenever a class is added or removed from a node – grows more expensive with a bigger DOM. Recycling your DOM nodes means that we are going to keep the total number of DOM nodes considerably lower, making all these processes faster.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: Does this really need to be parenthetical?

Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: Could DOM recycling use some example code?

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+The first hurdle is the scrolling itself. Since we will only have a tiny subset of all available items in the DOM at any given time, we need to find another way to make the browser’s scrollbar properly reflects the amount content that is theoretically there. We will use a 1px by 1px sentinel element with a transform to force the container element – the runway – to have the desired height. We have to be careful to keep the layer of the runway completely empty, though. No background color, nothing. The second the layer contains more than that the browser’s optimizations won’t skip processing that layer and we will have to store a texture on our graphics card that has a height of a couple of hundred thousand pixels. Definitely not viable on a mobile device.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: Runway - You should define your terms with a reference to your diagram before throwing them at your reader.

Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: I'm not following the bit about keeping the runway empty.

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+      <rect class="whitener" x="800" y="-600" width="5000" height="1800"/>
+      <g class="viewport">
+        <rect x="0" y="0" width="800" height="600" />
+        <text x="-600" y="-16" transform="rotate(-90)" font-size="72">Viewport</text>
+      </g>
+      </g>
+    </svg>
+
+The same goes for scrolling in the other direction. We will, however, never shrink the runway in our implementation, so that the scrollbar position stays consistent.
+Tombstones
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: Is this a heading?

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+<img src="/web/updates/images/2016/07/infinite-scroller/anchoring.png">
+
+After things in the runway have changed, we can now restore a situation that feels almost identical to the user. Win! Except a resized window means that each items has potentially changed its height, so how do we know how far down the anchored content should be placed? We don’t! To find out we would have to layout every element above the anchored item and add up all of their heights; this could cause a significant pause after a resize, and we don’t want that. Instead, we resort to assuming that every item above is the same size as a tombstone and adjust our scroll position accordingly. As elements are scrolled into the runway, we adjust or scroll position, effectively deferring the layout work to when it is actually needed.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: identical to the user? What does it feel identical too?

Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: 'As elements are scrolled into the runway' Do you mean as elements are scrolled into the viewport. 

'we adjust or scroll position' > 'we adjust our scroll position'.

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+
+#Layout
+I have skipped over an important detail: Layout. Each recycling of a DOM element would normally relayout the entire runway which would bring us well below our target of 60 frames per second. To avoid this, we are taking the burden of layout onto ourselves and use absolutely positioned elements with transforms. This way we can pretend that all the elements further up the runway are still taking up space when in actuality there is only empty space. Since we are doing layout ourselves, we can cache the positions each item ends up and we can immediately load the correct element from cache when the user scrolls backwards.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: 'we can cache the positions each item ends up' > 'we can cache the positions where each item ends up'

File: src/content/en/updates/posts/2016/07/infinite-scroller.md [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+
+#Bleeding-edge tweaks
+Recently, [Chrome added support for CSS Containment][Containment], a feature that allows us developers to tell the browser an element is a boundary for layout and paint work. Since we are doing layout ourselves here, it’s a prime application for containment. Whenever we add an element to the runway, we *know* the other items don’t need to be affected by the relayout. So each item should be get `contain: layout`. We also don‘t want to affect the rest of our website, so the runway itself should get this style directive as well.
+
+Another thing we considered is using [IntersectionObservers] as a mechanism to detect when the user has scrolled far enough for us to start recycling elements and load new data. However, IntersectionObservers are specified to be high latency (as if using `requestIdleCallback`), so we might actually *feel* less responsive with IntersectionObservers than without. Even our current implementation using the `scroll` event suffers from this problem, as scroll events are dispatched on a “best effort” basis. Eventually, [Houdini’s Compositor Worklet][Compositor Worklet] would be the high fidelity solution to this problem.
```
Comment (jpmedley) [inline src/content/en/updates/posts/2016/07/infinite-scroller.md]: I think part of the link is missing. Also IntersectionObserver should have backticks.

**Comment by jpmedley** [comment]
I found a few issues that need to be addressed.

**Comment by surma** [comment]
Implemented your review @jpmedley. Thank you very much.

Live preview here: http://pr-3043-dot-weasel-dot-web-central.appspot.com/web/updates/2016/07/infinite-scroller?hl=en

**Comment by surma** [comment]
@petele @jeffposnick Can we get this in for this weeks push?

**Comment by petele** [comment]
@surma - yep, will merge now and either push today or tomorrow (Tokyo time)

**Comment by oldergod** [comment]
@surma Just in case, the `IntersectionObservers` does not get hyperlinkyfied.

**Pull request merged**